base/compiler/ssair/inlining.jl

```diff
@@ -1392,22 +1396,33 @@ function compute_inlining_cases(@nospecialize(info::CallInfo), flag::UInt32, sig
             handled_all_cases &= handle_any_const_result!(cases,
                 result, match, argtypes, info, flag, state; allow_typevars=true)
         end
+        if !fully_covered
+            atype = argtypes_to_type(sig.argtypes)
+            # We will emit an inline MethodError so we need a backedge to the MethodTable
+            unwrapped_info = info isa ConstCallInfo ? info.call : info
+            if unwrapped_info isa UnionSplitInfo
+                for (fullmatch, mt) in zip(unwrapped_info.fullmatches, unwrapped_info.mts)
+                    !fullmatch && push!(state.edges, mt, atype)
+                end
+            elseif unwrapped_info isa MethodMatchInfo
+                push!(state.edges, unwrapped_info.mt, atype)
+            else @assert false end
```

Comment:
This assertion does not hold in cases where external `CallInfo` is used, e.g. the test case in the following diff causes this code path to be hit:
```diff
diff --git a/test/compiler/AbstractInterpreter.jl b/test/compiler/AbstractInterpreter.jl
index d95354cefa..6c745218c0 100644
--- a/test/compiler/AbstractInterpreter.jl
+++ b/test/compiler/AbstractInterpreter.jl
@@ -431,6 +431,8 @@ end
 @inline function inlined_usually(x, y, z)
     return x * y + z
 end
+foo_split(x::Float64) = 1
+foo_split(x::Int) = 2
 
 # check if the inlining algorithm works as expected
 let src = code_typed1((Float64,Float64,Float64)) do x, y, z
@@ -444,6 +446,7 @@ let NoinlineModule = Module()
     main_func(x, y, z) = inlined_usually(x, y, z)
     @eval NoinlineModule noinline_func(x, y, z) = $inlined_usually(x, y, z)
     @eval OtherModule other_func(x, y, z) = $inlined_usually(x, y, z)
+    @eval NoinlineModule bar_split_error() = $foo_split(Core.compilerbarrier(:type, nothing))
 
     interp = NoinlineInterpreter(Set((NoinlineModule,)))
 
@@ -473,6 +476,11 @@ let NoinlineModule = Module()
         @test count(isinvoke(:inlined_usually), src.code) == 0
         @test count(iscall((src, inlined_usually)), src.code) == 0
     end
+
+    let src = code_typed1(NoinlineModule.bar_split_error)
+        @test count(iscall((src, foo_split)), src.code) == 0
+        @test count(iscall((src, Core.throw_methoderror)), src.code) > 0
+    end
 end
 
 # Make sure that Core.Compiler has enough NamedTuple infrastructure
```

To handle such cases, we probably need to introduce a new `CallInfo` interface `getmatchinfo(::CallInfo, idx::Int)::Tuple{#=fullmatch=#Bool, #=mt=#MethodTable}` and enforce its implementation for each `CallInfo`, and use it here:
```suggestion
            for i = 1:nunion
                fullmatch, mt = getmatchinfo(info, i)
                !fullmatch && push!(state.edges, mt, atype)
            end
```

Reply:
Oh, that's a very good catch (and a useful test!) thanks.

That change looks pretty easy - I'll add it to the PR.

Still, I'm curious - It seems like `abstract_applicable` / `find_matching_methods` do not support custom `CallInfo`, etc. Is that because those are modeling a function application via the global method table, and therefore we expect the "standard" CallInfo semantics?

Reply:
> Still, I'm curious - It seems like `abstract_applicable` / `find_matching_methods` do not support custom `CallInfo`, etc. Is that because those are modeling a function application via the global method table, and therefore we expect the "standard" CallInfo semantics?

In fact, `find_matching_methods` is meant to retrieve the results of method matching, and it is not intended to create `CallInfo`, especially `MethodMatchInfo` or `UnionSplitInfo`. Currently, these objects are created within `find_matching_methods` for convenience, but I guess it might be more sensible for the caller side, such as `abstract_call_gf_by_type` (which calls `find_matching_methods`), to create those `CallInfo` objects. Custom `CallInfo` is used when an external `AbstractInterpreter` injects some logic into those routines, typically to trigger special behavior such as inlining.